diff -urN a/runopts.h b/runopts.h
--- a/runopts.h	2022-11-14 11:15:12
+++ b/runopts.h	2023-02-27 13:10:11
@@ -135,6 +135,8 @@
 	/* points into pubkey_plugin */
 	char *pubkey_plugin_options;
 #endif
+	char *shell;
+	char *path;
 
 	int pass_on_env;
 
diff -urN a/svr-chansession.c b/svr-chansession.c
--- a/svr-chansession.c	2022-11-14 11:15:12
+++ b/svr-chansession.c	2023-02-27 13:10:21
@@ -1012,7 +1012,9 @@
 	addnewvar("LOGNAME", ses.authstate.pw_name);
 	addnewvar("HOME", ses.authstate.pw_dir);
 	addnewvar("SHELL", get_user_shell());
-	if (getuid() == 0) {
+	if (svr_opts.path != NULL) {
+		addnewvar("PATH", svr_opts.path);
+	} else if (getuid() == 0) {
 		addnewvar("PATH", DEFAULT_ROOT_PATH);
 	} else {
 		addnewvar("PATH", DEFAULT_PATH);
@@ -1021,6 +1023,7 @@
 		addnewvar("LANG", cp);
 		m_free(cp);
 	}	
+
 	if (chansess->term != NULL) {
 		addnewvar("TERM", chansess->term);
 	}
diff -urN a/svr-runopts.c b/svr-runopts.c
--- a/svr-runopts.c	2022-11-14 11:15:12
+++ b/svr-runopts.c	2023-02-27 13:10:03
@@ -178,6 +178,8 @@
         svr_opts.pubkey_plugin = NULL;
         svr_opts.pubkey_plugin_options = NULL;
 #endif
+	svr_opts.shell = NULL;
+	svr_opts.path = NULL;
 	svr_opts.pass_on_env = 0;
 	svr_opts.reexec_childpipe = -1;
 
@@ -328,6 +330,12 @@
 					break;
 				case 'z':
 					opts.disable_ip_tos = 1;
+					break;
+				case 'S':
+					next = &svr_opts.shell;
+					break;
+				case 'H':
+					next = &svr_opts.path;
 					break;
 				default:
 					fprintf(stderr, "Invalid option -%c\n", c);
